Initial support for Fibaro HomeCenter hubs

pbalogh77 opened the pull request:
## Description:
Added Fibaro component to integrate with Fibaro Home Center, using an adapted version of the Fiblary library and the Fibaro REST API.
What works:
- Lights, Switches, Sensors, Covers, Binary sensors
- Connection, updates, interaction

What's not working yet:
- Virtual devices, Scenes, Panels, Locks, etc

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):**
https://github.com/home-assistant/home-assistant.io/pull/7493

## Example entry for `configuration.yaml` (if applicable):
```yaml
fibaro:
    url: "http://yourfibarohc/api/"
    username: "your@superuseremail.com"
    password: "YourPassword1"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pbalogh77 committed:
Fibaro HC connection, initial commit

Very first steps working, connects, fetches devices, represents sensors, binary_sensors and lights towards HA.

pbalogh77 committed:
Cover, switch, bugfixes

Initial support for covers
Initial support for switches
Bugfixes

pbalogh77 committed:
Some cleanup and improved lights

pylint based cleanup
light switches handled properly
light features reported correctly

pbalogh77 committed:
Added status updates and actions

Lights, Blinds, Switches are mostly working now

pbalogh77 committed:
Code cleanup, fiblary3 req

Fiblary3 is now in pypi, set it as req
Cleanup based on pylint

pbalogh77 committed:
Included in .coveragerc and added how to use guide

Included the fibaro component in coveragerc
Added usage instructions to file header

pbalogh77 committed:
PyLint inspired fixes

Fixed pylint warnings

pbalogh77 committed:
PyLint inspired fixes

PyLint inspired fixes

pbalogh77 committed:
updated to fiblary3 0.1.5

frenck commented:
Could not find a related PR in our documentation repository. Adding `docs-missing` label.

pbalogh77 committed:
Minor fixes to finally pass pull req

Fixed fiblary3 to work with python 3.5
Updated fiblary3 to 0.1.6
(added energy and batteryLevel dummies)

pbalogh77 commented:
Could someone please help me understand what I'm doing wrong here? I keep getting the this error from lint: https://travis-ci.org/home-assistant/home-assistant/jobs/447813324
fiblary3 is listed as a requirement, it's in requirements_all.txt and all works fine locally, even the req validate script.

pbalogh77 committed:
module import and flake8 fixes

Finally (hopefully) figured out what lint is complaining about

pbalogh77 commented:
What does the coverage/coveralls error mean and what can I do about it? My component is listed in .coveragerc omit section.

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,414 @@
+from collections import defaultdict
+import threading
+import voluptuous as vol
+from fiblary3.client.v4.client import Client as FibaroClient
```
Thanks, I fixed it.

pbalogh77 committed:
Fixed color support for lights, simplified callback

Fixed color support for lights
Simplified callback for updates
Uses updated fiblary3 for color light handling

pbalogh77 committed:
Lean and mean refactor

While waiting for a brave reviewer, I've been making the code smaller and easier to understand.

pbalogh77 committed:
Minor fixes to please HoundCI

pbalogh77 committed:
Removed unused component

Scenes are not implemented yet

pbalogh77 committed:
Nicer comments.

pbalogh77 committed:
DEVICE_CLASS, ignore plugins, improved mapping

Added support for device class and icons in sensors and binary_sensors
Improved mapping of sensors and added heuristic matching
Added support for hidden devices
Fixed conversion to float in sensors

pbalogh77 committed:
Fixed dimming

Fibaro apparently does not need, nor like the extra turnOn commands for dimmers

pbalogh77 committed:
flake8

pbalogh77 committed:
Cleanup, Light fixes, switch power

Cleanup of the component to separate init from connect, handle connection error better
Improved light handling, especially for RGBW strips and working around Fibaro quirks
Added energy and power reporting to switches

pbalogh77 committed:
Missing comment added

Missing comment added to please flake8

pbalogh77 commented:
Could someone please take a look at the error message here? I don't think it's related to my code in any way.

pbalogh77 committed:
Removed everything but bin.sensors

Stripdown, hoping for a review

pbalogh77 committed:
better aligned comments

OMG

amelchio commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+            if power:
+                return convert(power, float, 0.0)
+        else:
+            return 0
```
`return None` rather than zero?

amelchio replied:
It feels like `current_power_w` should not be in the base class.

amelchio replied:
Agreed that code copying should be avoided. I was thinking that you could have a `FibaroPowerDevice` and not have to decide capabilities dynamically.

Anyway, let's just keep it as is for now.

amelchio commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+                    int(self.fibaro_device.properties.batteryLevel)
+            if 'fibaroAlarmArm' in self.fibaro_device.interfaces:
+                attr[ATTR_ARMED] = True if \
+                    self.fibaro_device.properties.armed else False
```
`attr[ATTR_ARMED] = bool(self.fibaro_device.properties.armed)`

amelchio replied:
Actually, you can even avoid the `bool()` cast if `armed` is itself a boolean :)

pbalogh77 replied:
My trouble is that I haven't fully tested this functionality, so I'll need to test and see to be sure

amelchio commented on homeassistant/components/binary_sensor/fibaro.py:
```diff
@@ -0,0 +1,72 @@
+    """Representation of a Fibaro Binary Sensor."""
+
+    _icon = None
+    _device_class = None
```
Why not move these inside `__init__` ?

pbalogh77 replied:
You mean the declarations and initializations? I'm coming from c/c++ and somehow it feels nicer to know what variables the functions of a class use, takes a bit of the guesswork out. Not sure if it's pythonic :-) If you say it should be moved, I'm happy to move it, tho.

amelchio replied:
Python does not use explicit declarations, what you have here are class variables (static members in C++). It is not exactly wrong because your instance variables will override them ... but it looks odd.

The guesswork is largely avoided by initializing all instance variables in the `__init__` method.

pbalogh77 replied:
I see. Thanks. I'll move it in a future PR, if that's okay

amelchio replied:
I would prefer that we do this now so we have the pattern right when the remaining device types arrive.

It seems like we should always end up with a not-None `stype`? In that case these initializers can just be removed.

pbalogh77 replied:
stype can be None. I couldn't find a documentation of all the various device types and basetypes Fibaro HC supports and can report, so these are more like heuristics, trying to match it with known types, but at the same type keeping any other type support. Even if their icon might not be specialized, but at least all binary sensors would be supported this way.
I did some cleanup, to only initialize with None if needed.

amelchio replied:
Right, I forgot that devices are also added based on capabilities.

amelchio reviewed (commented):
This already looks good to me. I have a bunch of small questions/proposals but I did not find any major issues.

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+        self._read_devices()
+        return True
+
+    def __del__(self):
```
I've no idea on how my class gets destructed, I'll look at the event

amelchio replied:
Are you still looking at the event or did you determine that it does not need to be called at all?

pbalogh77 replied:
I'll get back to it in a future PR

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+                    _LOGGER.warning("Error updating %s data of %s, not found",
+                                    property_name,
+                                    self._device_map[device_id].ha_id)
+                if self._callbacks.get(device_id, None):
```
yes, it'd be neater, I'll clean it up

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+        callback_set = set()
+        for change in state.get('changes', []):
+            device_id = change.pop('id')
+            for property_name, value in list(change.items()):
```
Do you mean  "for change in state.get('changes', []):"?
I've copy-pasted that line for Fiblary's source code and did not try to change it.

amelchio replied:
I meant `list(change.items())`, the `list` seems to needlessly create a copy (but I could have missed something).

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+        self._room_map = {}
+        for room in rooms:
+            self._room_map[room.id] = room
+        return True
```
Thanks, never done this, but seen it elsewhere. I'll adapt it, it looks cleaner and more compact.

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+        self._room_map = {}
+        for room in rooms:
+            self._room_map[room.id] = room
+        return True
```
I plan on later implementing error handling pathways, not sure whether through return values or exceptions yet. I was also not sure if there'd be anything here that could fail. Finally there's nothing really, so you're right.

pbalogh77 committed:
Fixes based on code review

Fixes based on code review

amelchio commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,360 @@
+    # 'sensor',
+    # 'light',
+    # 'switch',
+    # 'cover',
```
Yeah, I know. Still, for this PR it is mostly noise :)

pbalogh77 replied:
Removed.

pbalogh77 committed:
Implemented stopping

Implemented stopping of StateHandler thread
Cleanup for clarity

pbalogh77 committed:
Minor fix

Removed unnecessary list copying

amelchio commented on homeassistant/components/fibaro.py:
```diff
@@ -210,12 +211,20 @@ def setup(hass, config):
+
+    def stop_fibaro(event):
+        """Stop Fibaro Thread."""
+        _LOGGER.info("Shutting down Fibaro Hub")
```
I think this message is slightly excessive since it is basically just a progress log.

pbalogh77 replied:
I agree, but I just followed exactly what the Xiaomi and the Vera components are doing and saying.

amelchio replied:
I see. If you agree, just remove it – you should be able to find examples of that as well :-)

pbalogh77 committed:
Nicer wording on shutdown

amelchio reviewed (approved):
Great, I think we can go with this.

MartinHjelmare commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+            for property_name, value in change.items():
+                if property_name == 'log':
+                    if value and value != "transfer OK":
+                        _LOGGER.info("LOG %s: %s",
```
Max debug.

pbalogh77 replied:
ok

MartinHjelmare commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+                    _LOGGER.debug("%s (%s, %s) not mapped",
+                                  device.ha_id, device.type,
+                                  device.baseType)
+        return True
```
Nothing is checking this return value.

pbalogh77 commented on homeassistant/components/binary_sensor/fibaro.py:
```diff
@@ -0,0 +1,71 @@
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Perform the setup for Fibaro controller devices."""
+    add_entities(
```
If this component is not yet using discovery, so I have no use for the discovery_info parameter, do I still need to check for it and return if None?

MartinHjelmare replied:
Yes. Users can add binary_sensor platform config sections which will try to set up the platform from config. That's what we are guarding from with the guard clause. We only want to set up the platform if loaded from component via discovery.

pbalogh77 replied:
ok, thanks, I fixed it

pbalogh77 commented on homeassistant/components/binary_sensor/fibaro.py:
```diff
@@ -0,0 +1,71 @@
+    def __init__(self, fibaro_device, controller):
+        """Initialize the binary_sensor."""
+        self._state = None
+        FibaroDevice.__init__(self, fibaro_device, controller)
```
Since it has two parent classes, I'm not sure your proposal is okay, since the two parents have different parameters for their init. (BinarySensor class has none, FibaroDevice has these parameter needs).

pbalogh77 replied:
Ok, I changed it. I'm coming from C/C++ and it gives me a bad feeling overall, as this feels like less defensive code, but I also see why it makes sense.

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+            _LOGGER.error('Error while udpating attributes for %s',
+                          self.ha_id)
+
+        attr['Id'] = self.ha_id
```
I copy-pasted this part from Vera and Xiaomi. I'll change it nevertheless.

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+    def _update_callback(self):
+        """Update the state."""
+        self.update()
+        self.schedule_update_ha_state(False)
```
Sorry, it's a left-over from debugging, fixing it

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+        self.controller = controller
+        self._name = fibaro_device.friendly_name
+        self.ha_id = fibaro_device.ha_id
+        self.controller.register(fibaro_device.id, self._update_callback)
```
ok

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+                attr[ATTR_CURRENT_ENERGY_KWH] = convert(
+                    self.fibaro_device.properties.energy, float, 0.0)
+        except (ValueError, KeyError):
+            _LOGGER.error('Error while udpating attributes for %s',
```
ok

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+
+    def stop_fibaro(event):
+        """Stop Fibaro Thread."""
+        _LOGGER.info("Shutting down Fibaro connection.")
```
ok

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+
+import logging
+from collections import defaultdict
+import voluptuous as vol
```
ok

pbalogh77 commented on homeassistant/components/fibaro.py:
```diff
@@ -0,0 +1,351 @@
+"""
+Support for the Fibaro devices.
+
+To enable, add the following section to configuration.yaml:
```
ok

pbalogh77 commented on homeassistant/components/binary_sensor/fibaro.py:
```diff
@@ -0,0 +1,71 @@
+        FibaroDevice.__init__(self, fibaro_device, controller)
+        self.entity_id = ENTITY_ID_FORMAT.format(self.ha_id)
+        stype = None
+        if fibaro_device.type in SENSOR_TYPES.keys():
```
ok

pbalogh77 committed:
Minor changes based on code review

pbalogh77 committed:
minor fixes based on code review

MartinHjelmare reviewed (approved):
Looks good! Can be merged when build passes.

pbalogh77 committed:
removed extra line break

amelchio commented:
This was an awesome first contribution. The code is high quality and your response to feedback is exemplary.
